explanation/performance/perf-pgo.md

```diff
@@ -0,0 +1,153 @@
+```
+
+On an AMD system, use the following invocation:
+
+```bash
+$ sudo perf record \
+	-e 'cpu/event=0xc4,umask=0x0,name=ex_ret_brn_tkn/' \
+	--branch-any \
+	--all-cpus \
+	--output openssl-nonpgo.perfdata \
+	-- openssl speed -mr -seconds 60 -evp md5 sha512 rsa2048 aes-256-cbc
+```
+
+Note that the only thing that differs is the PMU event to be monitored.
+
+After the command has finished running, you should see a file named `openssl-nonpgo.perfdata` in your current directory.  This is the file we will now need to convert to `gcov`.
+
+### Converting `perfdata` to `gcov` with `autofdo`
+
+With `autofdo` installed, you can convert the `perfdata` generated in the last step to `gcov` by doing:
+
+```bash
+$ sudo chown ubuntu:ubuntu openssl-nonpgo.perfdata
+$ create_gcov \
+	--binary /usr/bin/openssl \
+	--gcov openssl-nonpgo.gcov \
+	--profile openssl-nonpgo.perfdata \
+	--gcov_version 2 \
+	-use_lbr false
+```
+
+`create_gcov` is verbose and will display several messages that may look like something is wrong; you should inspect its output and also make sure to check the `$?` shell variable to make sure that it has finished successfully.  You will then have a file called `openssl-nonpgo.gcov` in your current directory, which is the `gcov` file we will use to feed back to GCC when recompiling OpenSSL.  You can inspect the file by using the `dump_gcov` tool.
+
+### Rebuilding OpenSSL with PGO
+
+We have everything we need to rebuild OpenSSL and make use of our profile data.  The most important thing now is to set the `CFLAGS` environment variable properly so that GCC can find and use the profile we generated in the previous step.
+
+How you perform this `CFLAGS` adjustment depends on how you built the software in the first place, so we won't cover this part here.  The resulting `CFLAGS` variable should have the following GCC option, though:
+
+```
+-fauto-profile=/path/to/openssl-nonpgo.gcov
+```
+
+Make sure to adjust the path to the `openssl-nonpgo.gcov` file.
+
+Also note that enabling PGO will make GCC automatically enable several optimization passes that are usually disable, even when using `-O2`.  This can lead to new warnings/errors in the code, which might require either the code to be adjusted or the use of extra compilation flags to suppress these warnings.
+
+After the software is built, you can install this new version and run the benchmark tests again to compare results.
+
+### Our results
+
+We obtained some interesting results while performing the tests.  Here are the highlights:
+
+* Overall, the performance gain/loss was minimal.  Most of the time it was less than 1%.
+
+* There were some huge outliers, though.  For example, `sha512` showed a gain of more than 13% in one of the runs, but also showed a loss of 11% in another run!  This is likely noise due to external factors, even though we tried to isolate our test environment as much as possible.  Still, it was surprising to see such discrepancies.
```

Comment:
the noise as an excuse is not really helpful, but we can make it helpful.
For example we can show "but by disabling this and that in background" and "by running it N times excluding the outliers" and ... we have found this to be ...
And then, as you did well in touching but not explaining all elude to more like "but this is about geting reliable statistical data out of measurements which shall not be the topic here in all its detail"

Reply:
It could be helpful to suggest an order of magnitude number of tests that could be needed to provide a decent signal-to-noise ratio. I assume you'll be reporting the number of tests you ran in your blog post, but still I think it's worth reporting on what could be considered a statistically useful number of tests

Reply:
@cpaelzer and @s-makin, thanks for the suggestions.

@cpaelzer, would you have some time to look deeper into this with me?  I'd like your input regarding the statistical side of things, especially when it comes to determining a good number of tests to be run.

Unfortunately I have already released the machine I was using for testing this, so if more tests are needed they will take some time.

Reply:
thanks for the discussion - go for:
1. positive tone
2. explain that non-improvement can happen even if doing right
3. hint at any general statistic trick to help like what sally and I mentioned
4. link/hint at blog post for an example of the same with improvements

Reply:
Thanks @cpaelzer .  I've updated the last sections, please take a look when you can.